paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java

```diff
@@ -0,0 +1,116 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+
+/**
+ * This class is a parallel execution util class, which mainly aim to process tasks parallelly with
+ * memory control.
+ */
+public class ParallellyExecuteUtils {
+
+    // reduce memory usage by batch iterable process, the cached result in memory will be 2 *
+    // queueSize
+    public static <T, U> Iterable<T> parallelismBatchIterable(
+            Function<List<U>, List<T>> processor, List<U> input, int queueSize) {
+        if (queueSize <= 0) {
+            throw new NegativeArraySizeException("queueSize need to be positive");
+        }
+
+        Queue<List<U>> stack = new ArrayDeque<>();
+        int size = input.size();
+        int num = size / queueSize;
+
+        for (int i = 0; i < num; i++) {
+            stack.add(input.subList(i * queueSize, (i + 1) * queueSize));
+        }
+
+        if (num * queueSize < size) {
+            stack.add(input.subList(num * queueSize, size));
+        }
+
+        return () ->
+                new Iterator<T>() {
+                    CompletableFuture<List<T>> batch = null;
+                    List<T> activeList = null;
+                    private int index = 0;
+                    private T next = null;
+
+                    // build the first batch
+                    {
+                        if (stack.size() > 0) {
+                            batch =
+                                    CompletableFuture.supplyAsync(
+                                            () -> processor.apply(stack.poll()),
+                                            FileUtils.COMMON_IO_FORK_JOIN_POOL);
+                        }
+                    }
+
+                    @Override
+                    public boolean hasNext() {
+                        try {
+                            // get active list for polling, and submit tasks meanwhile
+                            if (activeList == null) {
+                                if (batch == null) {
+                                    return false;
+                                }
+
+                                activeList = batch.get();
+                                if (stack.size() > 0) {
+                                    batch =
+                                            CompletableFuture.supplyAsync(
+                                                    () -> processor.apply(stack.poll()),
+                                                    FileUtils.COMMON_IO_FORK_JOIN_POOL);
+                                } else {
+                                    batch = null;
+                                }
+                            }
+                            if (index < activeList.size()) {
+                                next = activeList.get(index++);
+                                if (index == activeList.size()) {
+                                    activeList = null;
```

Comment:
You implement a wrong iterator.
Test should cover multiple invoking for `hasNext `.

Reply:
done this. added tests for multiple invoking hasNext method